### PR TITLE
[Bugfix][DCP] Fix NVIDIA DeepSeek-V3.2 / GLM-5.2 fused attention

### LINKED ISSUE #50095: [Bug][DCP] NVIDIA DeepSeek-V3.2 / GLM-5.2 fused attention bypasses DCP handling
# [Bug][DCP] DeepSeek-V3.2 / GLM-5.2 fused attention bypasses pure-DCP handling

## Problem

The fused DeepSeek-V3.2 / GLM-5.2 path still has two pure-DCP correctness
defects:

1. `vllm/models/deepseek_v32/common/kernels.py` checks owner-local KV slot
   validity before query RMSNorm. A negative slot means another DCP rank owns
   the token's KV entry; it does not remove this rank's query contribution.
2. `vllm/models/deepseek_v32/attention.py` gathers query heads and combines
   partial output/LSE only for PCP+DCP. Pure DCP therefore sends rank-local
   query heads into attention over owner-local KV and never performs the
   required cross-rank result merge.

The sparse-indexer DCP rank, world-size, and interleave plumbing mentioned in
the original report has landed independently on `main`; it is no longer part
of this issue.

## Observable regression

A focused kernel reproducer uses an all-negative slot mapping, representing a
DCP rank that owns no KV-cache slot for those tokens.

Expected:

- Query output matches reference RMSNorm.
- MLA and indexer KV caches remain unchanged.
- The Top-K buffer is cleared.

Actual on unpatched `main`:

- Query output remains zero because the kernel returns before query RMSNorm.

At model level, @drakosha reproduced repeated-token output on unpatched
GLM-5.2-NVFP4 with TP4/DCP4 and `fp8_ds_mla`. The patched path produced output
byte-identical to DCP1 for the tested deterministic greedy request.

## Expected behavior

- Query RMSNorm runs independently of owner-local KV-slot validity.
- Negative slots continue to suppress MLA and indexer KV-cache writes.
- Pure DCP gathers query heads before owner-local sparse attention.
- Partial output/LSE is combined across DCP ranks through the existing DCP
  manager.

## Proposed fix and validation

PR #50005 implements the fix and regression coverage.

At its current head:

- The focused fused norm/RoPE suite passes all 67 tests on GB200.
- All applicable changed-file pre-commit hooks pass.
- Independent 4xH200 GLM-5.2-NVFP4 TP4/DCP4 validation passes the deterministic
  DCP1-versus-DCP4 comparison described above.

This correctness work is independent of the Shared-DCP performance feature.
OpenAI Codex assisted with investigation, validation orchestration, and
drafting; human review remains required.

### PR DESCRIPTION
**Pull request by foraxe**
# [Bugfix][DCP] Fix NVIDIA DeepSeek-V3.2 / GLM-5.2 fused attention

Fixes #50095.

## Purpose

Fix two DCP correctness defects in the fused DeepSeek-V3.2 / GLM-5.2
attention path:

- The fused norm/RoPE kernel skipped query RMSNorm when a DCP rank had no
  owner-local KV-cache slot. KV ownership does not remove that rank's query
  contribution.
- Pure DCP sent only rank-local query heads into attention over owner-local KV
  and did not combine the resulting partial output/LSE across DCP ranks.

The sparse-indexer DCP rank, world-size, and interleave plumbing described in
the original version of this PR has since landed independently on `main`; this
PR no longer changes that part.

## Implementation

- Run query RMSNorm before the owner-local slot check. Negative slots still
  suppress MLA and indexer KV-cache writes.
- For pure DCP, gather query heads through `MLADCPManager` before sparse MLA
  attention. The manager keeps the direct-workspace fast path when supported
  and falls back safely for oversized mixed batches.
- Combine partial output/LSE through the existing `MLADCPManager` after
  attention. For this NVIDIA pure-DCP mixed-batch path, pass no PCP-only decode
  metadata; its sparse backend supplies neutral contributions for empty local
  shards.
- Extend the existing fused norm/RoPE test with an all-negative-slot regression
  that checks valid query output, unchanged caches, and Top-K clearing.

## Non-duplication

I re-audited related open work on August 23, 2026. Searching for issue #50095
finds only this PR.

- #52472 adds a separate TurboQuant sparse MLA backend and does not modify this
  fused DeepSeek-V3.2 attention or norm/RoPE path.
- #52377 has landed with generic sparse-MLA forced-MQA metadata and
  direct-workspace fallback handling after the DCP-manager refactor; this PR
  now consumes that manager path but still supplies the NVIDIA fused override
  and negative-slot query-normalization fixes.
- #47779 enables DCP in the SM120 FlashInfer sparse MLA wrapper.
- #50009 is independent Shared-DCP performance work.

None fixes both live defects in
`vllm/models/deepseek_v32/{attention.py,common/kernels.py}`.

## Validation

Current upstream base inspected: `b2dd9ce73d`  
Current PR head: `22e9a19f8d`

The branch was forward-ported across the August DeepSeek-V3.2 file moves by
@drakosha in foraxe/vllm#1. Against current `main`, the PR contains only the
three intended files. `git diff --check` passes.

On one GB200 GPU at the exact PR head:

```bash
CUDA_VISIBLE_DEVICES=0 PYTHONPATH=$PWD \
  /workspace/vllm_cp/vllm/.venv/bin/python -m pytest \
  tests/kernels/test_fused_deepseek_v32_norm_rope.py -q
# 67 passed
```

All applicable changed-file pre-commit hooks pass, including Ruff, formatting,
mypy, SPDX, forbidden-import, and accelerator-API checks.

Independent forward-port validation by @drakosha used 4xH200 with
GLM-5.2-NVFP4, TP4/DCP4, `fp8_ds_mla`, and greedy decoding. On the tested
prompt, patched DCP4 output was byte-identical to DCP1; unpatched `main`
produced repeated-token output. The focused kernel suite passed 67 tests with
the fix, while the new negative-slot regression was the only failure without
it.

## AI assistance and human review

OpenAI Codex assisted with investigation, implementation, validation
orchestration, forward-port review, and drafting. @drakosha reports AI
assistance for the forward port and H200 validation.

- [x] The submitter has reviewed every changed line, verified the reported
  local evidence, and can explain and defend the query-gather and output/LSE
  merge protocol.

### CODE REVIEW

**Comment by foraxe** [comment]
Tracking issue: #50095.

#50005 is the proposed fix for the NVIDIA DeepSeek-V3.2 / GLM-5.2 fused path bypassing DCP query normalization, sparse-indexer geometry, query-head gathering, and partial output/LSE merging.

This correctness work is independent of the experimental Shared-DCP performance feature.

**Comment by drakosha** [comment]
@foraxe this fix is still needed. Both defects are live on `main` as of `da329cc30`: the early `return` on `slot_mapping < 0` in `common/kernels.py` still skips the `pid == 2` query RMSNorm, and the head gather plus cross-rank LSE merge in `attention.py` are still behind `use_pcp`. GLM-5.2 with plain DCP=4 produces repeated-token garbage on current main because of it.

The branch conflicts because the DSA path moved in August (`deepseek_v32/nvidia/{attention,kernels}.py` became `deepseek_v32/attention.py` and `common/kernels.py`). I brought your branch up to current main and opened https://github.com/foraxe/vllm/pull/1 against it. It is a merge commit, so it applies without conflicts; merging it there updates this PR and drops `needs-rebase`. The fix is unchanged, your regression test came with it, and the commit keeps you as author.

What we can add is validation on hardware, which this PR did not have: on 4xH200 with GLM-5.2-NVFP4 (TP4 + DCP4, `fp8_ds_mla`) the patched build gives DCP=4 output byte-identical to DCP=1 under greedy decoding. The kernel suite passes 67 tests with the fix; without it your new case is the only failure.

If you would rather hand this off, say so and I will open a superseding PR with you as author. Otherwise it is yours, and I will leave it a week before doing anything else with it.

AI assistance was used for the rebase and the runs.

**Comment by foraxe** [comment]
The forward-port from @drakosha has been reviewed and merged. Current head `ffeea6f3d5` is mergeable against current `main` and contains only the three intended files. On GB200, the focused fused norm/RoPE suite passes all 67 tests, and all applicable changed-file pre-commit hooks pass. I refreshed this PR and #50095 to the current two-defect scope; indexer DCP geometry has already landed independently. The visible pre-run-check failure is only the CI authorization gate because this PR does not yet have a maintainer-applied `ready` or `verified` label. @GirasoleY @WoosukKwon, since you recently worked on this refactored path, a review would be appreciated.

**Comment by cjackal** [comment]
It seems the changes in `vllm/models/deepseek_v32/attention.py` is a dsv32-localized version of #52377 (As commented in this PR, sparse MLA metadata has no `.decode` attribute, and sparse MLA mixed batches use `dcp_manager.query_gather` with long prefills, both of this and mine fixes by fallback to allgather).

I don't know model archs using sparse MLA other than DSV32/DSV4, so either change makes sense I guess, just want to mention it FYI.

**Comment by drakosha** [comment]
@cjackal we can settle the "either change" question with a run. On 4xH200 with GLM-5.2-NVFP4 (TP4 + DCP4, `fp8_ds_mla`): nightly `ba07e4a48` + #52377 alone still emits repeated-token garbage at DCP=4; nightly + this PR gives output byte-identical to DCP=1 under greedy.

They are not interchangeable because of routing: since #52861 those two archs go through `vllm/models/deepseek_v32/` on NVIDIA, so the call sites #52377 fixes are no longer executed there. Your PR still covers the generic path (ROCm keeps the compiled route), so both are needed, one per path.

Your `dcp_manager.query_gather` point matches what we hit independently: its workspace is decode-sized while a sparse mixed batch carries prefill rows, so the allgather fallback is right.

AI assistance was used for the runs.

**Comment by cjackal** [comment]
> They are not interchangeable because of routing: since #52861 those two archs go through `vllm/models/deepseek_v32/` on NVIDIA, so the call sites #52377 fixes are no longer executed there. Your PR still covers the generic path (ROCm keeps the compiled route), so both are needed, one per path.

Thank you for the pointer! I missed the recent MRV2 forward override indeed, my `.decode` patch no longer works unless users explicitly set `VLLM_USE_MODEL_RUNNER_V2=0`.

File: vllm/models/deepseek_v32/attention.py [hunk 1]
```diff
@@ -512,29 +512,54 @@ def _sparse_indexer_and_attn(
...
-            query_start_loc = attn_metadata.query_start_loc[
-                : attn_metadata.num_decodes + 1
-            ]
+            else:
+                # Sparse mixed batch: every row (decode and prefill) is a
+                # DCP-partial result, and the impl already neutralizes rows
+                # whose local shard holds none of the selected tokens with
+                # the (0, -inf) merge identity, so no seq_len-based
+                # empty-shard masking is needed (FlashMLASparseMetadata has
+                # no `decode` view to source it from either).
```
Comment (LucasWilkinson) [inline vllm/models/deepseek_v32/attention.py]: nit: can you please tighten up this comment?

File: vllm/models/deepseek_v32/attention.py [hunk 2]
```diff
@@ -512,29 +512,54 @@ def _sparse_indexer_and_attn(
             if isinstance(mqa_q_arg, tuple):
                 mqa_q_arg = torch.cat(mqa_q_arg, dim=-1)
             mqa_q_arg = get_tp_group().all_gather(mqa_q_arg, dim=1)
+        elif not self.use_pcp and self.impl.dcp_world_size > 1:
+            # Pure DCP: all-gather the query heads across the DCP group so
+            # every rank attends to its local KV shard with the full head
+            # set; the LSE merge below reduce-scatters the heads back.
+            # Plain all-gather rather than dcp_manager.query_gather: the
+            # sparse mixed batch also gathers prefill rows, which can exceed
+            # the decode-sized direct symmetric-memory workspace.
```
Comment (LucasWilkinson) [inline vllm/models/deepseek_v32/attention.py]: nit: can you please tighten up this comment and make a TODO to switch to `dcp_manager.query_gather`

@GirasoleY can we fix this in the `dcp_manager` ? maybe fallback to `all_gather` for oversized inputs? or make it so that we can flag support for mixed-batches to `dcp_manager` and it allocates appropriately sized workspaces?

**Review by LucasWilkinson — APPROVED** [review]
Thanks for the contribution! left a couple nits hopefully you can address before landing

File: vllm/models/deepseek_v32/attention.py (hunk 2 above)
Comment (cjackal) [reply vllm/models/deepseek_v32/attention.py]: Regarding the DCP symmem workspace size issues, I think the q-gather / combine fallbacks in my #52377 should be good in general.

File: vllm/models/deepseek_v32/attention.py [hunk 3]
```diff
@@ -512,29 +512,54 @@ def _sparse_indexer_and_attn(
...
+            assert self.dcp_manager is not None
+            if isinstance(mqa_q_arg, tuple):
+                mqa_q_arg = torch.cat(mqa_q_arg, dim=-1)
+            mqa_q_arg = self.dcp_manager.group.all_gather(mqa_q_arg, dim=1)
```
Comment (cjackal) [inline vllm/models/deepseek_v32/attention.py]: Wonder if the fallback approach in #52377 + `mqa_q_arg = self.dcp_manager.query_gather(mqa_q_arg)` is preferred over hardcoded allgather in model architecture script in general; I consider DCP Manager as the centralized communication controller for the lifetime of the model engine so it would be better fixing and direct calling DCP Manager than hardcoding the gathering method here in the long run I think.

Comment (cjackal) [reply vllm/models/deepseek_v32/attention.py]: Another gain in #52377 is that we can keep using the faster direct workspace path for pure-decode batches, I might add.

File: vllm/models/deepseek_v32/attention.py (hunk 2 above)
Comment (LucasWilkinson) [reply vllm/models/deepseek_v32/attention.py]: Agreed; I mostly just wanted to get the fix in but since https://github.com/vllm-project/vllm/pull/52377 looks GTG lets just land that first then update this PR

File: vllm/models/deepseek_v32/attention.py (hunk 1 above)
Comment (foraxe) [reply vllm/models/deepseek_v32/attention.py]: Tightened this to the two-line empty-shard invariant in `6127a179f8`. The focused GB200 suite remains 67/67 and the changed-file pre-commit hooks pass.

File: vllm/models/deepseek_v32/attention.py (hunk 2 above)
Comment (foraxe) [reply vllm/models/deepseek_v32/attention.py]: Tightened the comment and added the requested `dcp_manager.query_gather` TODO in `6127a179f8`. Agreed on landing #52377 first; it is approved but still open, so I am not duplicating its manager fallback here. Once it lands, I will switch this call to the manager path and remove the TODO. Leaving this thread open until then.

File: vllm/models/deepseek_v32/attention.py (hunk 3 above)
Comment (foraxe) [reply vllm/models/deepseek_v32/attention.py]: Agreed. Centralizing this in `MLADCPManager` also preserves the direct workspace for pure decode. I will update this call after #52377 lands rather than copying its manager changes into this PR. Leaving the thread open for that follow-up.

Comment (LucasWilkinson) [reply vllm/models/deepseek_v32/attention.py]: https://github.com/vllm-project/vllm/pull/52377 has landed 👍

File: vllm/models/deepseek_v32/attention.py (hunk 2 above)
Comment (foraxe) [reply vllm/models/deepseek_v32/attention.py]: Updated now that #52377 has landed. In 22e9a19f8d I merged current main, removed the temporary TODO, and switched the GLM path to dcp_manager.query_gather. This keeps the direct-workspace path for supported shapes and delegates oversized-input fallback to MLADCPManager. Focused CUDA validation: 67 passed; changed-file pre-commit and mypy also passed.

File: vllm/models/deepseek_v32/attention.py (hunk 3 above)
Comment (foraxe) [reply vllm/models/deepseek_v32/attention.py]: Implemented in 22e9a19f8d after #52377 landed: the model now calls dcp_manager.query_gather instead of hardcoding group.all_gather, so communication selection remains centralized and pure-decode batches retain the direct-workspace fast path.

Comment (foraxe) [reply vllm/models/deepseek_v32/attention.py]: Hi @LucasWilkinson, I have updated this #50005 with the #52377, could you take a look?

**Comment by ehfd** [comment]
@LucasWilkinson Can we get this merged?

**Comment by foraxe** [comment]
#50005 merged. 
Thanks @LucasWilkinson @drakosha @cjackal @ehfd
